### ISSUE TITLE
Parameter-dependent decomposition rules cannot specialize on constant parameters under capture

### ISSUE THREAD

**Issue by Qottmann** [PennyLaneAI/catalyst#3172]
### Summary

Under `qml.qjit(capture=True)`, Catalyst's graph-based decomposition compiles a **single, generic decomposition rule per operator type**, traced with an **abstract** parameter, and applies it uniformly via the MLIR `decompose-lowering` pass. As a result, decomposition rules whose *structure* depends on the concrete value of a parameter cannot specialize, i.e. the parameter is always abstract at rule-compile time.

The concrete motivating case is the **adaptive-precision RZ phase-gradient decomposition** here: https://github.com/PennyLaneAI/pennylane/pull/10098. It truncates the `SemiAdder` width (and drops angles that round to zero) based on the concrete angle bits, guarded by:

```python
if adaptive_precision and not math.is_abstract(phi):
    ...  # truncate angle_wires / phase_grad_wires to the significant width
```

Under `capture=True`, `phi` is always abstract when the rule is compiled, so this branch is dead code: the adder is always full width and the Toffoli/T counts are identical for `adaptive_precision=True` and `False`. The large (~70%) Toffoli/T reduction that this rule produces under `capture=False` silently vanishes under capture.

### Why "the angles are constants" does not help

Even when the circuit has no arguments and the angles are compile-time constants, the concreteness is not exploited. The angles *are* concrete literals in the captured plxpr:

```python
import numpy as np, pennylane as qml, jax
qml.capture.enable()

angles = [f * 2 * np.pi for f in (1/2, 1/4, 3/4, 1/64, 7/8, 2.0**-7)]

def circuit():
    for a in angles:
        qml.RZ(a, 0)
    return qml.expval(qml.Z(0))

print(jax.make_jaxpr(circuit)())
# ... operator[op_cls=RZ ...] 3.141592653589793:f32[] 0:i32[]
# ... operator[op_cls=RZ ...] 1.5707963267948966:f32[] 0:i32[]
# ... operator[op_cls=RZ ...] 0.09817477042468103:f32[] 0:i32[]  (etc.)
```

But the decomposition rule is compiled **once**, generically, keyed by op-name + wire-count rather than by the concrete angle. The concrete value flows into the compiled rule as a runtime SSA operand (the emitted MLIR computes `binary_decimals(phi)` at "rule runtime"), so the per-angle structural truncation has nowhere to happen.

The rule still produces correct results, just without the optimization, so this is a silent performance regression relative to `capture=False`.

### Environment

- catalyst `0.16.0-dev86`
- pennylane `0.46.0-dev88`
- jax `0.7.1`
- macOS arm64

written with the help of genAI

**Comment by Qottmann** [comment]
### Minimal reproducer (toggle `CAPTURE`)

A custom `RX` rule that branches on `math.is_abstract(phi)`. Toggle `CAPTURE` and compare the rule's view of `phi`. No dependency on the phase-gradient rule, and it does **not** run the crashing `DecomposeLoweringPass` (#3171).

```python
import pennylane as qml
from pennylane import math
from pennylane.decomposition import register_resources

qml.decomposition.enable_graph()

CAPTURE = True  # <-- toggle this


@register_resources(lambda phi, wires: {qml.RZ: 1})
def rx_to_rz(phi, wires):
    print("  rule body: is_abstract(phi) =", math.is_abstract(phi))
    qml.RZ(phi, wires)


@qml.transforms.decompose(gate_set={"RZ", "GlobalPhase"}, fixed_decomps={qml.RX: rx_to_rz})
@qml.qnode(qml.device("null.qubit", wires=1))
def circuit():
    qml.RX(0.5, 0)  # compile-time-constant angle
    return qml.expval(qml.Z(0))


if CAPTURE:
    qml.qjit(circuit, capture=True, target="mlir").mlir
else:
    qml.specs(circuit)()
```

### Output

```
CAPTURE = False
  rule body: is_abstract(phi) = False          # applied with the concrete angle

CAPTURE = True
  rule body: is_abstract(phi) = False          # PennyLane graph-solve
  rule body: is_abstract(phi) = True           # <-- Catalyst compiles the applied rule with abstract phi
```

Under capture the rule that the pass actually applies is compiled with an abstract `phi` (in the emitted MLIR, `phi` is the SSA arg `%arg1: tensor<f64>`), so any `math.is_abstract`-guarded specialization is dead — even though `0.5` is a concrete literal in the plxpr. This is why adaptive-precision truncation / zero-angle elision silently no-op under capture.

**Comment by kipawaa** [comment]
This has been discussed before - what is "dynamic" data and what is "static" data should vary per _rule_, not per _operator_ as it does now. This would require some significant changes throughout the stack, though, and may or may not be compatible with the current systems.

**Comment by paul0403** [comment]
> The angles are concrete literals in the captured plxpr:
> ```python
> ... operator[op_cls=RZ ...] 3.141592653589793:f32[] 0:i32[]
> ```

Note that this isn't just a problem in decomposition. Jax only prints these as constants as an illusion. In reality the number is a `ConstVar`, which is considered an SSA value. True static constants on an jaxpr equation would be a param (aka in the `[]` square brackets, like `op_cls` here).

**Comment by paul0403** [comment]
From a decomp point of view, a rule subroutine function simply does not know the concrete value of its arguments when it is being compiled. The arguments must be staged out as tracers. PennyLane graph-solve doesn't have this problem because the rules are not being compiled as subroutines. 

If a rule needs an arg as static, it needs that arg to be registered as an arg in the `compilable_argnames` of the operator it's on.

**Comment by paul0403** [comment]
See https://github.com/PennyLaneAI/catalyst/blob/40c1f9163fa1d6032862aae4c802d0f21527f500/frontend/test/lit/test_operator2/test_decomposition_rule_entry_point.py#L267

this test that is compiling a rule whose body is using the static value of an argument

**Comment by Qottmann** [comment]
we will definitely need to be able to do this in the future, utilizing the concrete value of a parameter can lead to huge savings in quantum resources
